---
title: "Internal QC"
output: 
  html_document:
    toc: true
    toc_depth: 2
date: "`r Sys.Date()`"
params:
  merged_rds: ""
  metrics_csv: ""
---
<style type="text/css">
.main-container {
  max-width: 100% !important;
  margin: auto;
}
</style>

```{r setup, include=F}
options(width = 1800)
knitr::opts_chunk$set(echo = FALSE, message=F, warning=F)
```

```{r libraries, include=F}
suppressMessages(library(Seurat))
suppressMessages(library(plotly))
suppressMessages(library(tidyverse))
suppressMessages(library(patchwork))
suppressMessages(library(randomcoloR))
#plan("multiprocess", workers = detectCores())
```

```{r testdata2, eval=F}
seuratobj=readRDS('/mnt/preqc/LargeData/samples_merged_ds.rds')
md_combined=seuratobj@meta.data 
md_doublet=slice_sample(md_combined, prop=0.1)
merged_metrics=read_csv('/mnt/preqc/LargeData/metrics_merged.csv', show_col_types = FALSE)
```

```{r inputdata_test, eval=F}
seuratobj <- readRDS('/mnt/10x_endtoend_pipeline/LargeData/scallop.rds')
merged_metrics=read_csv('/mnt/10x_endtoend_pipeline/LargeData/metrics_merged.csv')
md_combined=seuratobj@meta.data
md_doublet=seuratobj@meta.data %>% filter(doublet_class=='doublet')
```

```{r inputdata, eval=T}
seuratobj <- params$merged_rds
merged_metrics=params$metrics_csv
md_combined=seuratobj@meta.data
md_doublet=seuratobj@meta.data %>% filter(doublet_class=='doublet')
```


```{r figdimensions}
#Number of rows
subplot_dim=300
sample_count=length(unique(md_combined$orig.ident))
cols_number=ifelse(sample_count<5, sample_count, 5)

rows_number=ceiling(sample_count/5)
rows_number_vln=ceiling(sample_count/25)
figure_width=cols_number*subplot_dim

figure_height1=rows_number*subplot_dim+200 #For ncount vs nfeature
figure_height2=rows_number*subplot_dim+100 #For mitochonrial
figure_height_vln=rows_number_vln*subplot_dim+250 #For mitochonrial
figure_umap_height=rows_number*subplot_dim # For umap without sliders
if (sample_count>25){
  head_rows_no=round(sample_count/rows_number_vln)
  last_row_no=sample_count-(rows_number_vln-1)*head_rows_no
vlnplot_batch_df=data.frame(orig.ident=sort(unique(md_combined$orig.ident)), 
                            vlnplot_batch=c(rep(seq(1:(rows_number_vln-1)), each=head_rows_no), 
                                            rep(rows_number_vln, last_row_no)))
md_combined=merge(md_combined, vlnplot_batch_df)
subplot_vlnwidth=head_rows_no*-3.3+153.3
figure_width_vln=head_rows_no*subplot_vlnwidth
} else {
  md_combined$vlnplot_batch=1
  subplot_vlnwidth=sample_count*-3.3+153.3
  figure_width_vln=sample_count*subplot_vlnwidth
}


```


<style>

.RNAthreshold {
  position:relative;
  height: 650px;
  width: `r figure_width`;
  overflow-x:auto;
  overflow-y:auto
}
.mitothreshold {
  position:relative;
  height: 650px;
  width: `r figure_width`;
  overflow-x:auto;
  overflow-y:auto
}
.primarymetrics {
  position:relative;
  height: 500px;
  width: `r figure_width`;
  overflow-x:auto;
  overflow-y:auto
}
</style>

## Primary metrics {.tabset}
Primary metrics generated by cellranger.
```{r gexmetrics, fig.width=800/72, fig.height=sample_count*50/72, dpi=72}
assays=unique(merged_metrics$`Library Type`)
gexsamples=5 #Need a value for code chunk fig size evaluation, regardless of whether eval is true or not
bcrsamples=5
tcrsamples=5
adtsamples=5
gex_metrics=merged_metrics %>% filter(`Library or Sample`=='Sample', 
                          `Library Type`=='Gene Expression', 
                          `Metric Name` %in% c('Cells', 
                                                'Median UMI counts per cell',
                                                'Median genes per cell',
                                                'Median reads per cell'
                                                )) %>%
  distinct(Sample, `Library Type`, `Metric Name`, .keep_all = T)%>% #Somehow the metrics are repeated
  mutate(`Metric Value`=readr::parse_number(`Metric Value`))
gexsamples=length(unique(gex_metrics$Sample)) #Used for determining figure height
primary_div1 <- knitr::asis_output('<div class="primarymetrics">')
primary_div2 <- knitr::asis_output('</div>')

```

```{r gexmetrics_plot, eval='Gene Expression' %in% assays, fig.width=800/72, fig.height=(gexsamples*15+50)/72*4, dpi=72, class.output="primarymetrics"}
maxheight=(gexsamples*15+50)/72*4
p=ggplot(gex_metrics, aes(y=Sample, x=`Metric Value`, label=`Metric Value`))+
  geom_col(fill='#99d8c9')+facet_wrap(`Metric Name`~., ncol = 1, scales = 'free')+
  ggtitle('GEX metrics')+geom_text(hjust="inward")
knitr::asis_output('### GEX\nMetrics for gene expression assay')
primary_div1
p
primary_div2
```

```{r adtmetrics, eval='Antibody Capture' %in% assays}
adt_metrics=merged_metrics %>% filter(`Library or Sample`=='Sample', 
                          `Library Type`=='Antibody Capture', 
                          `Metric Name` %in% c('Antibody reads usable per cell', 
                                                'Median UMI counts per cell'
                                                )) %>%
  distinct(Sample, `Library Type`, `Metric Name`, .keep_all = T)%>% #Somehow the metrics are repeated
  mutate(`Metric Value`=readr::parse_number(`Metric Value`))
adtsamples=length(unique(adt_metrics$Sample)) #Used for determining figure height
```

```{r adtmetrics_plot, eval='Antibody Capture' %in% assays, fig.width=800/72, fig.height=(adtsamples*15+60)/72*2, dpi=72}
maxheight=max(maxheight, (adtsamples*15+50)/72*2)
p=ggplot(adt_metrics, aes(y=Sample, x=`Metric Value`, label=`Metric Value`))+
  geom_col(fill='#99d8c9')+facet_wrap(`Metric Name`~., ncol = 1, scales = 'free')+
  ggtitle('ADT metrics')+geom_text(hjust="inward")
knitr::asis_output('### ADT\nMetrics for ADT assay')
primary_div1
p
primary_div2
```


```{r bcrmetrics, eval='VDJ B' %in% assays}
bcr_metrics=merged_metrics %>% filter(`Library or Sample`=='Sample', 
                          `Library Type`=='VDJ B', 
                          `Metric Name` %in% c('Estimated number of cells',
                                               'Median IGH UMIs per Cell',
                                               'Median IGK UMIs per Cell',
                                               'Median IGL UMIs per Cell',
                                               'Number of cells with productive V-J spanning pair'
                                                )) %>%
  distinct(Sample, `Library Type`, `Metric Name`, .keep_all = T)%>% #Somehow the metrics are repeated
  mutate(`Metric Value`=readr::parse_number(`Metric Value`))
bcrsamples=length(unique(bcr_metrics$Sample)) #Used for determining figure height
```

```{r bcrmetrics_plot, eval='VDJ B' %in% assays, fig.width=800/72, fig.height=(bcrsamples*15+50)/72*5, dpi=72 }
maxheight=max(maxheight, (bcrsamples*15+50)/72*5)
p=ggplot(bcr_metrics, aes(y=Sample, x=`Metric Value`, label=`Metric Value`))+
  geom_col(fill='#99d8c9')+facet_wrap(`Metric Name`~., ncol = 1, scales = 'free')+
  ggtitle('BCR metrics')+geom_text(hjust="inward")
knitr::asis_output('### BCR\nMetrics for BCR assay')
primary_div1
p
primary_div2
```

```{r tcrmetrics, eval='VDJ T' %in% assays}
tcr_metrics=merged_metrics %>% filter(`Library or Sample`=='Sample', 
                          `Library Type`=='VDJ T', 
                          `Metric Name` %in% c('Estimated number of cells', 
                                                'Median TRA UMIs per Cell',
                                                'Median TRB UMIs per Cell',
                                                'Number of cells with productive V-J spanning pair'
                                                )) %>%
  distinct(Sample, `Library Type`, `Metric Name`, .keep_all = T)%>% #Somehow the metrics are repeated
  mutate(`Metric Value`=readr::parse_number(`Metric Value`))
tcrsamples=length(unique(tcr_metrics$Sample)) #Used for determining figure height
```

```{r tcrmetrics_plot, eval='VDJ T' %in% assays, fig.width=800/72, fig.height=(tcrsamples*15+50)/72*4, dpi=72}
maxheight=max(maxheight, (tcrsamples*15+50)/72*4)
p=ggplot(tcr_metrics, aes(y=Sample, x=`Metric Value`, label=`Metric Value`))+
  geom_col(fill='#99d8c9')+facet_wrap(`Metric Name`~., ncol = 1, scales = 'free')+
  ggtitle('TCR metrics')+geom_text(hjust="inward")
knitr::asis_output('### TCR\nMetrics for TCR assay')
primary_div1
p
primary_div2
```

## RNA thresholds {.tabset}
Plots to assist in setting thresholds for nCount and nFeature for RNA.
<button type="button" id="rnathresholdbutton">Toggle frame</button>
```{r slider_boundaries}
#Boundaries for nCount_RNA cutoff slider
nCmax_per=0.8
max_ncount=max(md_combined$nCount_RNA)
minmax_ncount=md_combined %>% group_by(orig.ident) %>% 
  slice_max(order_by=nCount_RNA, n=1) %>% 
  ungroup() %>% slice_min(order_by=nCount_RNA, n=1) %>%
  select(nCount_RNA) %>% unlist
nCmax_lb=floor(quantile(md_combined$nCount_RNA, nCmax_per)/1000)*1000
nCmax_ub=ceiling(max_ncount/1000)*1000

#Boundaries for nFeature_RNA cutoff slider
max_nfeature=max(md_combined$nFeature_RNA)
minmax_nfeature=md_combined %>% group_by(orig.ident) %>% 
  slice_max(order_by=nFeature_RNA, n=1) %>% 
  ungroup() %>% slice_min(order_by=nFeature_RNA, n=1) %>%
  select(nFeature_RNA) %>% unlist
nFmax_lb=floor(quantile(md_combined$nFeature_RNA, nCmax_per)/100)*100
nFmax_ub=ceiling(max_nfeature/100)*100

#Zoom dataset
zoomed_x=quantile(md_combined$nCount_RNA, probs=0.5)
zoomed_y=quantile(md_combined$nFeature_RNA, probs=0.5)
md_zoomed=md_combined %>% filter(nCount_RNA<zoomed_x, nFeature_RNA<zoomed_y)

#Mito
max_mito=max(md_combined$percent.mt)
```

```{r makeshapes}
#Initial line for nCount_RNA cutoff
shapes <-
  list(
    type = "rect",
    y0 = 0,
    y1 = nFmax_lb,
    x0 = 0,
    x1 = nCmax_lb,
    line = list(color = 'green', dash="dot")
  )

#Initial line for nCount_RNA lower cutoff
shapes_lower <-
  list(
    type = "line",
    y0 = 0,
    y1 = 0,
    xref='paper',
    x0 = 0,
    x1 = 1,
    line = list(color = 'green', dash="dot")
  )

#Initial line for nCount upper cutoff vln
shapes_nCount_vln <-
  list(
    type = "line",
    y0 = nCmax_lb,
    y1 = nCmax_lb,
    xref='paper',
    x0 = 0,
    x1 = 1,
    line = list(color = 'green', dash="dot")
  )

#Initial line for nFeature box vln
shapes_nFeatures_vln <-
  list(
    type = "rect",
    y0 = 100,
    y1 = nFmax_lb,
    xref='paper',
    x0 = 0,
    x1 = 1,
    line = list(color = 'green', dash="dot")
  )


#Make discrete color scale based on Blackbody
colorscale_values=seq.int(0.,1, length.out=7)
tickvals=(colorscale_values[2:7]-diff(colorscale_values[2:3])/2)*5
ticktext=c(seq(0,4),'>=5')
colorscale_colors=c('rgb(0, 0, 0)',  
                    'rgb(184, 0, 0)',  
                    'rgb(230, 126, 0)', 
                    'rgb(240, 228, 102)', 
                    'rgb(236, 244, 255)', 
                    'rgb(160, 200, 255)')
colorscale_index=rep(seq(6), each=2)
colorscale=mapply(list, 
                  c(0, rep(colorscale_values[2:6], each=2), 1), 
                  colorscale_colors[colorscale_index],
                  SIMPLIFY = F)

```

```{r makeRNAplots}
do_base=. %>%
  plot_ly(x = ~nCount_RNA, y = ~nFeature_RNA, width=subplot_dim, height=subplot_dim) %>%
  add_histogram2d(xbins=list(start=0, size=500), ybins=list(start=0, size=100),
                       coloraxis = "coloraxis", bingroup=1, hoverinfo='none') %>%
  layout(shapes = list(shapes))

do_scatterplot=. %>% 
  add_annotations(
    text = ~unique(orig.ident),
    x = 0,
    y = 1.00,
    yref = "paper",
    xref = "paper",
    xanchor = "left",
    yanchor = "top",
    showarrow = FALSE,
    font = list(color = 'white')
  ) %>%
  layout(xaxis = list(range=c(0,max_ncount), tickformat = "~s"),
         yaxis = list(range=c(0,max_nfeature), tickformat = "~s"))

do_dbl=. %>% 
  add_annotations(
    text = ~unique(orig.ident),
    x = 0,
    y = 1.00,
    yref = "paper",
    xref = "paper",
    xanchor = "left",
    yanchor = "top",
    showarrow = FALSE,
    font = list(color = 'white')
  ) %>%
  layout(xaxis = list(range=c(0,max_ncount), tickformat = "~s"),
         yaxis = list(range=c(0,max_nfeature), tickformat = "~s"))

do_zoomed=. %>%
  plot_ly(x = ~nCount_RNA, y = ~nFeature_RNA, width=subplot_dim, height=subplot_dim) %>%
  add_histogram2d(xbins=list(start=0, size=50), ybins=list(start=0, size=10),
                       coloraxis = "coloraxis", bingroup=2, hoverinfo='none') %>%
  layout(shapes = list(shapes_lower)) %>% 
  add_annotations(
    text = ~unique(orig.ident),
    x = 0,
    y = 1.00,
    yref = "paper",
    xref = "paper",
    xanchor = "left",
    yanchor = "top",
    showarrow = FALSE,
    font = list(color = 'white')
  ) %>%
  layout(xaxis = list(range=c(0,zoomed_x), tickformat = "~s"),
         yaxis = list(range=c(0,zoomed_y), tickformat = "~s"))

colmap=distinctColorPalette(length(unique(seuratobj$orig.ident)))
names(colmap)=sort(unique(seuratobj$orig.ident))

do_vln= function(data_df, column, shapes, ymax){
  p=plot_ly(
    data=data_df,
    y = as.formula(paste('~',column)),
    x=~orig.ident,
    split = ~orig.ident,
    color=~orig.ident,
    colors=colmap,
    type = 'violin',
    hoverinfo='none',
    meanline = list(visible = T),
     showlegend =F) %>%
    layout(shapes = list(shapes),
           yaxis = list(range=c(0,ymax), tickformat = "~s"),
           xaxis=list(visible=T))
}

scatter_list=md_combined %>%
  group_by(orig.ident) %>%
  do(p=do_base(.) %>% do_scatterplot)

dbl_list=md_doublet %>%
  group_by(orig.ident) %>%
  do(p=do_base(.) %>% do_dbl)

count_vln_list=md_combined %>%
  group_by(vlnplot_batch) %>%
  do(p=do_vln(., 'nCount_RNA', shapes_nCount_vln, max_ncount))

feature_vln_list=md_combined %>%
  group_by(vlnplot_batch) %>%
  do(p=do_vln(., 'nFeature_RNA', shapes_nFeatures_vln, max_nfeature))

zoomed_list=md_zoomed %>%
  group_by(orig.ident) %>%
  do(p=do_zoomed(.))


fig_scatter = subplot(scatter_list, nrows = rows_number, shareX = T, shareY = T, titleX =F, titleY=F,  margin =c(0.002, 0.002, 0.002,0.002))
fig_dbl=subplot(dbl_list, nrows = rows_number, shareX = T, shareY = T, titleX =F, titleY=F,  margin =c(0.002, 0.002, 0.002,0.002))
fig_zoomed = subplot(zoomed_list, nrows = rows_number, shareX = T, shareY = T, titleX =F, titleY=F,  margin =c(0.002, 0.002, 0.002,0.002))
fig_count_vln = subplot(count_vln_list, nrows = rows_number_vln, shareX = F, shareY = T, titleX =T, titleY=T,  margin =c(0.002, 0.002, 0.002,0.002))  
fig_feature_vln = subplot(feature_vln_list, nrows = rows_number_vln, shareX = F, shareY = T, titleX =T, titleY=T,  margin =c(0.002, 0.002, 0.002,0.002))  
```


```{r makesliders}
sample_index=seq(0, sample_count-1)
vln_row_index=seq(0, rows_number_vln-1)

xcord <- seq(nCmax_lb,nCmax_ub, by=1000)
ncount_slider <- list()
for (i in 1:length(xcord)) {  
  args=as.list(rep(xcord[i], sample_count))
  args_name=sprintf('shapes[%s].x1', c(sample_index))
  names(args)=args_name
  ncount_slider[[i]] <- list(method = "relayout", 
                       args = list(args),
                       label = xcord[i]) 
} 

ycord <- seq(nFmax_lb,nFmax_ub, by=100)
nfeature_slider <- list()
for (i in 1:length(ycord)) {  
  args=as.list(rep(ycord[i], sample_count))
  args_name=sprintf('shapes[%s].y1', c(sample_index))
  names(args)=args_name
  nfeature_slider[[i]] <- list(method = "relayout", 
                       args = list(args),
                       label = ycord[i]) 
} 

ycord_zoomed <- seq(0,zoomed_y, by=100)
nfeature_min_slider <- list()
for (i in 1:length(ycord_zoomed)) {
  args=as.list(rep(ycord_zoomed[i], sample_count*2))
  args_name=c(sprintf('shapes[%s].y0',sample_index),  sprintf('shapes[%s].y1', sample_index))
  names(args)=args_name
  nfeature_min_slider[[i]] <- list(method = "relayout",
                       args = list(args),
                       label = ycord_zoomed[i])
}

xcord <- seq(nCmax_lb,nCmax_ub, by=1000)
ncount_vln_slider <- list()
for (i in 1:length(xcord)) {  
  args=as.list(rep(xcord[i], rows_number_vln*2))
  args_name=c(sprintf('shapes[%s].y0',vln_row_index), sprintf('shapes[%s].y1', vln_row_index))
  names(args)=args_name
  ncount_vln_slider[[i]] <- list(method = "relayout", 
                       args = list(args),
                       label = xcord[i]) 
} 

ycord <- seq(nFmax_lb,nFmax_ub, by=100)
nfeature_upper_vln_slider <- list()
for (i in 1:length(ycord)) {  
  args=as.list(rep(ycord[i], rows_number_vln))
  args_name=sprintf('shapes[%s].y1', vln_row_index)
  names(args)=args_name
  nfeature_upper_vln_slider[[i]] <- list(method = "relayout", 
                       args = list(args),
                       label = ycord[i]) 
} 

ycord_zoomed <- seq(0,zoomed_y, by=100)
nfeature_lower_vln_slider <- list()
for (i in 1:length(ycord_zoomed)) {
  args=as.list(rep(ycord_zoomed[i], rows_number_vln))
  args_name=sprintf('shapes[%s].y0', vln_row_index)
  names(args)=args_name
  nfeature_lower_vln_slider[[i]] <- list(method = "relayout",
                       args = list(args),
                       label = ycord_zoomed[i])
}


# add slider control to plot
scatter_layout = . %>% 
  layout(autosize = F, width =  max(figure_width, 400), height = figure_height1,
         annotations = list(
                list(x = 0 , y = 0.5, text = "nFeature",
                     textangle = 270, xshift=-50,
                     showarrow = F, xref='paper', yref='paper', font=list(size=20)),
                list(x = 0.5 , y = 0, text = "nCount",
                     yshift=-50,
                     showarrow = F, xref='paper', yref='paper', font=list(size=20))
                ),
         coloraxis=list(colorscale=colorscale,  cmax=5, cmin=0, reversescale=F,
                        colorbar=list(x=1.02, y=1, yanchor='top', tickvals=tickvals, ticktext=ticktext, len=200, lenmode='pixels',
                                      thickness=20))) %>%
  config(modeBarButtonsToRemove = c('hoverClosestCartesian','hoverCompareCartesian', 'zoomIn','zoomOut','autoScale'),
         displaylogo=F
         )

scatter_slider= . %>%
  layout(sliders = list(
    list(
      active =0, 
      currentvalue = list(prefix = "nCount < "), 
      pad = list(t = 80),
      #y=-0.5,
      steps = ncount_slider,
      len=300,
      lenmode='pixels'),
    list(
      active = 0, 
      currentvalue = list(prefix = "nFeature < "), 
      pad = list(t = 180),
      #y=-1,
      steps = nfeature_slider,
      len=300,
      lenmode='pixels')
    )) 
  
fig_scatter <- fig_scatter %>% 
  scatter_layout %>%
  scatter_slider
  
fig_dbl <- fig_dbl %>% 
  scatter_layout %>%
  scatter_slider

fig_zoomed <- fig_zoomed %>% 
  layout(autosize = F, width = max(figure_width, 400), height = figure_height2,
         annotations = list(
                list(x = 0 , y = 0.5, text = "nFeature",
                     textangle = 270, xshift=-50,
                     showarrow = F, xref='paper', yref='paper', font=list(size=20)),
                list(x = 0.5 , y = 0, text = "nCount",
                     yshift=-50,
                     showarrow = F, xref='paper', yref='paper', font=list(size=20))
                ),
         coloraxis=list(colorscale=colorscale,  cmax=5, cmin=0, reversescale=F,
                        colorbar=list(x=1.02, y=1, yanchor='top', tickvals=tickvals, ticktext=ticktext, len=200, lenmode='pixels',
                                      thickness=20))) %>%
  config(modeBarButtonsToRemove = c('hoverClosestCartesian','hoverCompareCartesian', 'zoomIn','zoomOut','autoScale'),
         displaylogo=F
         ) %>%
  layout(sliders = list( list(
    active = 0, 
    currentvalue = list(prefix = "nFeature > "), 
    pad = list(t = 80),
    steps = nfeature_min_slider,
    len=300,
    lenmode='pixels')
    )) 

fig_count_vln=fig_count_vln %>%
  layout(autosize = F, width = max(figure_width_vln, 400), 
         height = figure_height_vln,
         sliders = list(
           list(
             active = 0, 
             currentvalue = list(prefix = "nCount < "), 
             pad = list(t = 80),
             steps = ncount_vln_slider,
             len=300,
             lenmode='pixels')) )%>%
  config(modeBarButtonsToRemove = c('hoverClosestCartesian','hoverCompareCartesian', 'zoomIn','zoomOut','autoScale'),
         displaylogo=F)

fig_feature_vln=fig_feature_vln %>%
  layout(autosize = F, width = max(figure_width_vln, 400), 
         height = figure_height_vln+100,
         sliders = list(
           list(
             active = 0,
             currentvalue = list(prefix = "nFeature < "),
             pad = list(t = 80),
             steps = nfeature_upper_vln_slider,
             len=300,
             lenmode='pixels'),
           list(
             active = 0, 
             currentvalue = list(prefix = "nFeature > "), 
             pad = list(t = 180),
             steps = nfeature_lower_vln_slider,
             len=300,
             lenmode='pixels')))%>%
  config(modeBarButtonsToRemove = c('hoverClosestCartesian','hoverCompareCartesian', 'zoomIn','zoomOut','autoScale'),
         displaylogo=F)

```

### nFeatures vs nCounts
Sliders at the bottom for visualizing cutoffs. 2D histogram of nFeature vs nCount. Bin size is 100 for nFeature and 500 for nCount. 
```{r}
htmltools::div(fig_scatter, class='RNAthreshold')
```

### Doublets
Sliders at the bottom for visualizing cutoffs. 2D histogram of doublets detected using scDblFinder. 
```{r}
htmltools::div(fig_dbl, class='RNAthreshold')
```

### nFeatures vs nCounts (zoomed) 
Sliders at the bottom for visualizing cutoff. 2D histogram of nFeature vs nCount for only the bottom 50 percentile of both variables to assist in picking the lower threshold for nFeature. The bin-size is 50 for nCount and 10 for nFeature. 
```{r}
htmltools::div(fig_zoomed, class='RNAthreshold')
```

### nCount violin
Sliders at the bottom for visualizing cutoff. 
```{r}
htmltools::div(fig_count_vln, class='RNAthreshold')
```

### nFeatures violin
Sliders at the bottom for visualizing cutoff.
```{r}
htmltools::div(fig_feature_vln, class='RNAthreshold')
```

## Mitochondrial threshold {.tabset}
<button type="button" id="mitobutton">Toggle frame</button>
```{r mitochondrialplotly, eval=T}
# Prepare line
shapes_mito <-
  list(
    type = "line",
    y0 = 2,
    y1 = 2,
    xref='paper',
    x0 = 0,
    x1 = 1,
    line = list(color = 'green', dash="dot")
  )

# Prepare plotting functions
do_mito_count=. %>%
   plot_ly(x = ~nCount_RNA, y = ~percent.mt, width=subplot_dim, height=subplot_dim) %>%
  add_histogram2d(xbins=list(start=0, size=500), ybins=list(start=0, size=1),
                       coloraxis = "coloraxis", bingroup=3, hoverinfo='none') %>%
  add_annotations(
    text = ~unique(orig.ident),
    x = 0.5,
    y = 1,
    yref = "paper",
    xref = "paper",
    xanchor = "left",
    yanchor = "top",
    showarrow = FALSE,
    font = list(color = 'white')
  ) %>%
  layout(shapes = list(shapes_mito),
         xaxis = list(range=c(0,max_ncount), tickformat = "~s"),
         yaxis = list(range=c(0,max_mito)))

# Generate plots
mito_count_list=md_combined %>%
  group_by(orig.ident) %>%
  do(p=do_mito_count(.))

mito_vln_list=md_combined %>%
  group_by(vlnplot_batch) %>%
  do(p=do_vln(., 'percent.mt', shapes_mito, max_mito))

# Assemble subplots
fig_mito_count = subplot(mito_count_list, nrows = rows_number, shareX = T, shareY = T, titleX =T, titleY=T,  margin =c(0.002, 0.002, 0.002,0.002))
fig_mito_vln = subplot(mito_vln_list, nrows = rows_number_vln, shareX = F, shareY = T, titleX =T, titleY=T,  margin =c(0.002, 0.002, 0.002,0.002))  

#Prepare slider layers
ycord_vln <- seq(2,max_mito, by=2)

mito_scatter_slider <- list()
for (i in 1:length(ycord_vln)) {  
  args=as.list(rep(ycord_vln[i], sample_count*2))
  args_name=c(sprintf('shapes[%s].y0', sample_index), sprintf('shapes[%s].y1', sample_index))
  names(args)=args_name
  mito_scatter_slider[[i]] <- list(method = "relayout", 
                             args = list(args),
                             label = ycord_vln[i]) 
} 

mito_vln_slider <- list()
for (i in 1:length(ycord_vln)) {  
  args=as.list(rep(ycord_vln[i], rows_number_vln*2))
  args_name=c(sprintf('shapes[%s].y0', vln_row_index), sprintf('shapes[%s].y1', vln_row_index))
  names(args)=args_name
  mito_vln_slider[[i]] <- list(method = "relayout", 
                             args = list(args),
                             label = ycord_vln[i]) 
} 

# add slider control to plot
fig_mito_count <- fig_mito_count %>%
  layout(autosize = F, width = max(figure_width, 400), height = figure_height2,
         coloraxis=list(colorscale=colorscale,  cmax=5, cmin=0, reversescale=F,
                        colorbar=list(x=1.02, y=1, yanchor='top', tickvals=tickvals, ticktext=ticktext, len=200, lenmode='pixels',
                                      thickness=20)),
         sliders = list(
           list(
             active = 0, 
             currentvalue = list(prefix = "percent.mito < "), 
             pad = list(t = 50),
             steps = mito_scatter_slider,
             len=300,
             lenmode='pixels'))
         
         ) %>%
  config(modeBarButtonsToRemove = c('hoverClosestCartesian','hoverCompareCartesian', 'zoomIn','zoomOut','autoScale'),
         displaylogo=F
  )

fig_mito_vln <- fig_mito_vln %>%
  layout(autosize = F, width = max(figure_width_vln, 400), 
         height = figure_height_vln,
         sliders = list(
           list(
             active = 0, 
             currentvalue = list(prefix = "percent.mito < "), 
             pad = list(t = 100),
             steps = mito_vln_slider,
             len=300,
             lenmode='pixels')) )%>%
  config(modeBarButtonsToRemove = c('hoverClosestCartesian','hoverCompareCartesian', 'zoomIn','zoomOut','autoScale'),
         displaylogo=F
  )
```

### Percent mito vs UMI counts
Sliders at bottom for visualizing cut-off. 2D histogram of percentage mitochondrial RNA vs nCount. The bin size is 1 for percentage mitochondria RNA and 500 for nCount.
The slider bar at the bottom can be adjusted to visualize the percentage mitochondria RNA threshold.

```{r, eval=T}
htmltools::div(fig_mito_count, class='mitothreshold')
```

### Percent mito violin plot 
Sliders at bottom for visualizing cut-off. Violin plot of percentage mitochondrial RNA. 
```{r, eval=T}
htmltools::div(fig_mito_vln, class='mitothreshold')
```

## UMAPs {.tabset}
```{r}
umap_df=seuratobj[['umap']]@cell.embeddings %>% as.data.frame %>% 
  bind_cols(seuratobj@meta.data[,c('orig.ident','nCount_RNA','nFeature_RNA','percent.mt')])
plot_mapcell='mapcell_pbmc_RNA' %in% colnames(seuratobj@meta.data)
plot_WNN='predicted.WNN_PBMC' %in% colnames(seuratobj@meta.data)

get_short_name <- function(t){
  t <- ifelse(grepl('B ', t), 'B cells', t)
  t <- ifelse(grepl('CD4 ', t), 'CD4 T cells', t)
  t <- ifelse(grepl('CD8 ', t), 'CD8 T cells', t)
  t <- ifelse(grepl('cDC', t), 'cDC', t)
  t <- ifelse(grepl('NK', t), 'NK', t)
  t = ifelse(grepl('MAIT|ILC|HSPC|ASDC|Doublet|Platelet|Eryth', t), 'Others', t)
  t = ifelse(grepl('CD14 |CD16 ', t), 'Monocytes', t)
  t <- ifelse(grepl('Plasmablast', t), 'Plasma cells', t)
  return(t)
}

mapcell_agg=unlist(sapply(seuratobj@meta.data$mapcell_pbmc_RNA, get_short_name))
WNN_agg=unlist(sapply(seuratobj@meta.data$predicted.WNN_PBMC, get_short_name)) #Need to unlist to convert empty list to empty vector

celltypes_present=na.omit(unique(c(mapcell_agg, WNN_agg)))

#Make color map for cell annotation
if (length(celltypes_present)>0){
  colmap_celltype=DiscretePalette(length(celltypes_present))
  names(colmap_celltype)=sort(celltypes_present)}

```

### nCount
```{r,  fig.width=800/72, fig.height=700/72, dpi=72}
ggplot(umap_df, aes(x=umap_1, y=umap_2, color=nCount_RNA))+geom_point(size=0.5)+
  scale_colour_viridis_c(direction = -1)+ggtitle('nCount_RNA')
```

### nFeature
```{r, fig.width=800/72, fig.height=700/72, dpi=72}
ggplot(umap_df, aes(x=umap_1, y=umap_2, color=nFeature_RNA))+geom_point(size=0.5)+
  scale_colour_viridis_c(direction = -1)+ggtitle('nFeature_RNA')
```

### Percent.mt
```{r, fig.width=800/72, fig.height=700/72, dpi=72}
ggplot(umap_df, aes(x=umap_1, y=umap_2, color=percent.mt))+geom_point(size=0.5)+
  scale_colour_viridis_c(direction = -1)+ggtitle('Percent mitochondrial RNA')
```

### SampleID
```{r, fig.width=800/72, fig.height=700/72, dpi=72}
ggplot(umap_df, aes(x=umap_1, y=umap_2, color=orig.ident))+geom_point(size=0.5)+
  theme(legend.title=element_blank())+ scale_colour_manual(values = colmap)+ 
  guides(colour = guide_legend(override.aes = list(size=3)))+
  ggtitle('SampleIDs')
```

```{r, results='asis', eval=plot_mapcell}
cat('### Mapcell_PBMC')
```

```{r, eval=plot_mapcell, fig.width=800/72, fig.height=700/72, dpi=72}
umap_df=bind_cols(umap_df, mapcell_pbmc_agg=mapcell_agg)
ggplot(umap_df, aes(x=umap_1, y=umap_2, color=mapcell_pbmc_agg))+geom_point(size=0.5)+
   scale_colour_manual(values = colmap_celltype)+ guides(colour = guide_legend(override.aes = list(size=3)))+
  theme(legend.title=element_blank())+
  ggtitle("Mapcell PBMC RNA only")
```

```{r, results='asis', eval=plot_WNN}
cat('### WNN PBMC')
```

```{r, eval=plot_WNN, fig.width=800/72, fig.height=700/72, dpi=72}
umap_df=bind_cols(umap_df, WNN_agg=WNN_agg) 
ggplot(umap_df, aes(x=umap_1, y=umap_2, color=WNN_agg))+geom_point(size=0.5)+
   scale_colour_manual(values = colmap_celltype)+ guides(colour = guide_legend(override.aes = list(size=3)))+
  theme(legend.title=element_blank())+
  ggtitle("WNN PBMC")
```

## Versions
`r R.version.string`<br>
Seurat version: `r packageVersion("Seurat")`<br>

<script>
$("#rnathresholdbutton").click(function() {
    if ( $('.RNAthreshold').height() != 650)
          $('.RNAthreshold').height(650);
    else
          $('.RNAthreshold').height(`r figure_height1`);
});
$("#mitobutton").click(function() {
    if ( $('.mitothreshold').height() != 650)
          $('.mitothreshold').height(650);
    else
          $('.mitothreshold').height(`r figure_height2`);
});

</script>

